lib/remap-functions/src/join.rs

```diff
@@ -0,0 +1,156 @@
+use remap::prelude::*;
+use std::borrow::Cow;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Join;
+
+impl Function for Join {
+    fn identifier(&self) -> &'static str {
+        "join"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::Array(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "separator",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: false,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+        let separator = arguments.optional("separator").map(Expr::boxed);
+
+        Ok(Box::new(JoinFn { value, separator }))
+    }
+}
+
+#[derive(Clone, Debug)]
+struct JoinFn {
+    value: Box<dyn Expression>,
+    separator: Option<Box<dyn Expression>>,
+}
+
+impl Expression for JoinFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let array = self.value.execute(state, object)?.try_array()?;
+
+        let string_vec = array
+            .iter()
+            .map(|s| s.try_bytes_utf8_lossy().map_err(Into::into))
+            .collect::<Result<Vec<Cow<'_, str>>>>()
+            .map_err(|_| "all array items must be strings")?;
+
+        let separator: String = self
+            .separator
+            .as_ref()
+            .map(|s| {
+                s.execute(state, object)
+                    .and_then(|v| Value::try_bytes(v).map_err(Into::into))
+            })
+            .transpose()?
+            .map(|s| String::from_utf8_lossy(&s).to_string())
+            .unwrap_or_else(|| "".into());
```

Comment:
This whole thing would be simpler (in my opinion) to just use `match` instead.

```rust
match self.separator { 
  Some(expr) => expr.execute(state, object)?.try_bytes_utf8_lossy()?.to_owned(),
  None => "".to_string()
}
```

The only reason we need that extra allocation to a string is to make both arms of the match the same. If you wanted to avoid the extra allocation (*which is not really that critical*), you could just do the `join` in each branch.